partiql-grammar/src/main/antlr/org/partiql/grammar/PartiQL.g4

```diff
@@ -0,0 +1,614 @@
+
+expr
+    : exprOr;
+
+exprOr
+    : lhs=exprOr OR rhs=exprAnd     # Or
+    | parent=exprAnd                # ExprOrBase
+    ;
+
+exprAnd
+    : lhs=exprAnd op=AND rhs=exprNot  # And
+    | parent=exprNot                  # ExprAndBase
+    ;
+
+exprNot
+    : <assoc=right> op=NOT rhs=exprNot  # Not
+    | parent=exprPredicate              # ExprNotBase
+    ;
+
+exprPredicate
+    : lhs=exprPredicate op=(LT_EQ|GT_EQ|ANGLE_LEFT|ANGLE_RIGHT|NEQ|EQ) rhs=mathOp00  # PredicateComparison
+    | lhs=exprPredicate IS NOT? type                                                 # PredicateIs
+    | lhs=exprPredicate NOT? IN PAREN_LEFT expr PAREN_RIGHT                          # PredicateIn
+    | lhs=exprPredicate NOT? IN rhs=mathOp00                                         # PredicateIn
+    | lhs=exprPredicate NOT? LIKE rhs=mathOp00 ( ESCAPE escape=expr )?               # PredicateLike
+    | lhs=exprPredicate NOT? BETWEEN lower=mathOp00 AND upper=mathOp00               # PredicateBetween
+    | parent=mathOp00                                                                # PredicateBase
+    ;
+
+mathOp00
+    : lhs=mathOp00 op=CONCAT rhs=mathOp01
+    | parent=mathOp01
+    ;
+
+mathOp01
+    : lhs=mathOp01 op=(PLUS|MINUS) rhs=mathOp02
+    | parent=mathOp02
+    ;
+
+mathOp02
+    : lhs=mathOp02 op=(PERCENT|ASTERISK|SLASH_FORWARD) rhs=valueExpr
+    | parent=valueExpr
+    ;
+
+valueExpr
+    : sign=(PLUS|MINUS) rhs=valueExpr
+    | parent=exprPrimary
+    ;
+
+exprPrimary
+    : cast                       # ExprPrimaryBase
+    | sequenceConstructor        # ExprPrimaryBase
+    | substring                  # ExprPrimaryBase
+    | canCast                    # ExprPrimaryBase
+    | canLosslessCast            # ExprPrimaryBase
+    | extract                    # ExprPrimaryBase
+    | coalesce                   # ExprPrimaryBase
+    | dateFunction               # ExprPrimaryBase
+    | aggregate                  # ExprPrimaryBase
+    | trimFunction               # ExprPrimaryBase
+    | functionCall               # ExprPrimaryBase
+    | nullIf                     # ExprPrimaryBase
+    | exprPrimary pathStep+      # ExprPrimaryPath
+    | caseExpr                   # ExprPrimaryBase
+    | valueList                  # ExprPrimaryBase
+    | values                     # ExprPrimaryBase
+    | exprTerm                   # ExprPrimaryBase
+    ;
+
+/**
+ *
+ * PRIMARY EXPRESSIONS
+ *
+ */
+
+exprTerm
+    : PAREN_LEFT query PAREN_RIGHT   # ExprTermWrappedQuery
+    | parameter                      # ExprTermBase
+    | varRefExpr                     # ExprTermBase
+    | literal                        # ExprTermBase
+    | collection                     # ExprTermBase
+    | tuple                          # ExprTermBase
+    ;
+
+nullIf
+    : NULLIF PAREN_LEFT expr COMMA expr PAREN_RIGHT;
+
+coalesce
+    : COALESCE PAREN_LEFT expr ( COMMA expr )* PAREN_RIGHT;
+
+caseExpr
+    : CASE case=expr? (WHEN when=expr THEN then=expr)+ (ELSE expr)? END;
```

Comment:
Perhaps label the "else" expr to clarify the visitor

Reply:
I see your skip 2, and raise you ANTLR list labels.

```antlr
caseExpr
    : CASE case=expr? (WHEN conds+=expr THEN thens+=expr)+ (ELSE else=expr)? END;
```

```kotlin
override fun visitCaseExpr(ctx: PartiQLParser.CaseExprContext) = PartiqlAst.build {
        val pairs = ctx.conds.indicies.map { i -> exprPair(
            visitExpr(ctx.conds(i)),
            visitExpr(ctx.thens(i))
        )}
        val elseExpr = if (ctx.else_ != null) visitExpr(ctx.else_) else null
        val caseMeta = ctx.CASE().getSourceMetaContainer()
        when (ctx.case_) {
            null -> searchedCase(exprPairList(exprPairList), elseExpr, metas = caseMeta)
            else -> simpleCase(visitExpr(ctx.case_), exprPairList(exprPairList), elseExpr, metas = caseMeta)
        }
}
```

Reply:
This is great. Really like this idea. Thanks !